### PR TITLE
Cosmos integration legacy / transaction broadcast

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: core/src/wallet/cosmos/CosmosLikeMessage.cpp [hunk 1]
```diff
@@ -272,6 +272,7 @@ rapidjson::Value CosmosLikeMessage::toJson(rapidjson::Document::AllocatorType& a
         addString(kProposalId, content.proposalId, jsonContent, allocator);
 
         // cosmos::MsgVote::option
+        // FIXME autogenerated to_string() returns incorrect case (returns UPPERCASE but should be CamelCase)
         auto option = api::to_string(content.option);
```
Comment (gagbo) [inline core/src/wallet/cosmos/CosmosLikeMessage.cpp]: You can probably fix that by using the static functions in `CosmosLikeConstants()`

Comment (ghost) [reply core/src/wallet/cosmos/CosmosLikeMessage.cpp]: Good idea, will do.

File: core/src/wallet/cosmos/explorers/GaiaCosmosLikeBlockchainExplorer.cpp [hunk 2]
```diff
@@ -262,10 +262,12 @@ namespace ledger {
...
-        Future<String> GaiaCosmosLikeBlockchainExplorer::pushTransaction(
-            const std::vector<uint8_t>& transaction) {
-            return Future<String>::failure(Exception(
-                api::ErrorCode::IMPLEMENTATION_IS_MISSING, "Push Transaction is unimplemented"));
+        Future<String> GaiaCosmosLikeBlockchainExplorer::pushTransaction(const std::vector<uint8_t>& transaction) {
+            return _http->POST("/txs", transaction)
+                        .json().template map<String>(_executionContext, [] (const HttpRequest::JsonResult& result) -> String {
+                            auto& json = *std::get<1>(result);
+                            return json["result"].GetString();
+                        });
```
Comment (gagbo) [inline core/src/wallet/cosmos/explorers/GaiaCosmosLikeBlockchainExplorer.cpp]: Can you add the `accept: application/json` header to be sure we get the correct type in the callback ?

File: core/test/cosmos/Fixtures.cpp [hunk 3]
```diff
@@ -151,8 +166,6 @@ namespace ledger {
                         }
 
                         void assertSameTransaction(const Transaction& txRef, const Transaction& txResult) {
-                                std::cerr << "Tx check !!!" << std::endl;
-
```
Comment (gagbo) [inline core/test/cosmos/Fixtures.cpp:155]: :+1:  thanks I forgot I put that here :)

File: core/test/cosmos/db_test.cpp [hunk 4]
```diff
@@ -45,25 +45,23 @@ class CosmosDBTests : public BaseFixture {
...
     std::shared_ptr<CosmosLikeAccount> account;
     std::shared_ptr<CosmosLikeWallet> wallet;
     setupTest(pool, account, wallet);
 
     std::chrono::system_clock::time_point timeRef = DateUtils::now();
 
-    Message msg = setupSendMessage(timeRef);
-    Transaction tx = setupTransaction(std::vector<Message>{ msg }, timeRef);
+    auto msg = setupSendMessage();
+    auto tx = setupTransactionResponse(std::vector<Message>{ msg }, timeRef);
```
Comment (gagbo) [inline core/test/cosmos/db_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/db_test.cpp [hunk 5]
```diff
@@ -92,15 +90,15 @@ TEST_F(CosmosDBTests, BasicDBTest) {
...
```
Comment (gagbo) [inline core/test/cosmos/db_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/db_test.cpp [hunk 6]
```diff
@@ -139,15 +137,15 @@ TEST_F(CosmosDBTests, OperationQueryTest) {
...
```
Comment (gagbo) [inline core/test/cosmos/db_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/db_test.cpp [hunk 7]
```diff
@@ -171,16 +169,16 @@ TEST_F(CosmosDBTests, UnsuportedMsgTypeTest) {
...
     setupTest(pool, account, wallet);
 
     std::chrono::system_clock::time_point timeRef = DateUtils::now();
 
-    Message msgSend = setupSendMessage(timeRef);
-    Message msgVote = setupVoteMessage(timeRef);
-    Transaction tx = setupTransaction(std::vector<Message>{ msgSend, msgVote }, timeRef);
+    auto msgSend = setupSendMessage();
+    auto msgVote = setupVoteMessage();
+    auto tx = setupTransactionResponse(std::vector<Message>{ msgSend, msgVote }, timeRef);
```
Comment (gagbo) [inline core/test/cosmos/db_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 8]
```diff
@@ -127,9 +160,9 @@ TEST(CosmosTransaction, ParseRawMsgDelegateTransaction) {
     EXPECT_EQ(tx->serialize(), strTx);
 }
 
-TEST(CosmosTransaction, ParseRawMsgUndelegateTransaction) {
+TEST_F(CosmosTransactionTest, ParseRawMsgUndelegateTransaction) {
     auto strTx = "{\"account_number\":\"6571\",\"chain_id\":\"cosmoshub-3\",\"fee\":{\"amount\":[{\"amount\":\"5000\",\"denom\":\"uatom\"}],\"gas\":\"200000\"},\"memo\":\"Sent from Ledger\",\"msgs\":[{\"type\":\"cosmos-sdk/MsgUndelegate\",\"value\":{\"amount\":{\"amount\":\"1000000\",\"denom\":\"uatom\"},\"delegator_address\":\"cosmos102hty0jv2s29lyc4u0tv97z9v298e24t3vwtpl\",\"validator_address\":\"cosmosvaloper1grgelyng2v6v3t8z87wu3sxgt9m5s03xfytvz7\"}}],\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 9]
```diff
@@ -111,9 +144,9 @@ TEST(CosmosTransaction, ParseRawSignedMsgSendTransaction) {
               // clang-format on
 }
 
-TEST(CosmosTransaction, ParseRawMsgDelegateTransaction) {
+TEST_F(CosmosTransactionTest, ParseRawMsgDelegateTransaction) {
     auto strTx = "{\"account_number\":\"6571\",\"chain_id\":\"cosmoshub-3\",\"fee\":{\"amount\":[{\"amount\":\"5000\",\"denom\":\"uatom\"}],\"gas\":\"200000\"},\"memo\":\"Sent from Ledger\",\"msgs\":[{\"type\":\"cosmos-sdk/MsgDelegate\",\"value\":{\"amount\":{\"amount\":\"1000000\",\"denom\":\"uatom\"},\"delegator_address\":\"cosmos102hty0jv2s29lyc4u0tv97z9v298e24t3vwtpl\",\"validator_address\":\"cosmosvaloper1grgelyng2v6v3t8z87wu3sxgt9m5s03xfytvz7\"}}],\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 10]
```diff
@@ -143,9 +176,9 @@ TEST(CosmosTransaction, ParseRawMsgUndelegateTransaction) {
     EXPECT_EQ(tx->serialize(), strTx);
 }
 
-TEST(CosmosTransaction, ParseRawMsgBeginRedelegateTransaction) {
+TEST_F(CosmosTransactionTest, ParseRawMsgBeginRedelegateTransaction) {
     auto strTx = "{\"account_number\":\"6571\",\"chain_id\":\"cosmoshub-3\",\"fee\":{\"amount\":[{\"amount\":\"5000\",\"denom\":\"uatom\"}],\"gas\":\"200000\"},\"memo\":\"Sent from Ledger\",\"msgs\":[{\"type\":\"cosmos-sdk/MsgBeginRedelegate\",\"value\":{\"amount\":{\"amount\":\"1000000\",\"denom\":\"uatom\"},\"delegator_address\":\"cosmos102hty0jv2s29lyc4u0tv97z9v298e24t3vwtpl\",\"validator_dst_address\":\"cosmosvaloper1sd4tl9aljmmezzudugs7zlaya7pg2895ws8tfs\",\"validator_src_address\":\"cosmosvaloper1grgelyng2v6v3t8z87wu3sxgt9m5s03xfytvz7\"}}],\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 11]
```diff
@@ -160,9 +193,9 @@ TEST(CosmosTransaction, ParseRawMsgBeginRedelegateTransaction) {
     EXPECT_EQ(tx->serialize(), strTx);
 }
 
-TEST(CosmosTransaction, ParseRawMsgSubmitProposalTransaction) {
+TEST_F(CosmosTransactionTest, ParseRawMsgSubmitProposalTransaction) {
     auto strTx = "{\"account_number\":\"6571\",\"chain_id\":\"cosmoshub-3\",\"fee\":{\"amount\":[{\"amount\":\"5000\",\"denom\":\"uatom\"}],\"gas\":\"200000\"},\"memo\":\"Sent from Ledger\",\"msgs\":[{\"type\":\"cosmos-sdk/MsgSubmitProposal\",\"value\":{\"content\":{\"description\":\"My awesome proposal\",\"title\":\"Test Proposal\",\"type\":\"Text\"},\"initial_deposit\":[{\"amount\":\"1000000\",\"denom\":\"uatom\"}],\"proposer\":\"cosmos102hty0jv2s29lyc4u0tv97z9v298e24t3vwtpl\"}}],\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 12]
```diff
@@ -178,9 +211,9 @@ TEST(CosmosTransaction, ParseRawMsgSubmitProposalTransaction) {
     EXPECT_EQ(tx->serialize(), strTx);
 }
 
-TEST(CosmosTransaction, ParseRawMsgVoteTransaction) {
+TEST_F(CosmosTransactionTest, ParseRawMsgVoteTransaction) {
     auto strTx = "{\"account_number\":\"6571\",\"chain_id\":\"cosmoshub-3\",\"fee\":{\"amount\":[{\"amount\":\"5000\",\"denom\":\"uatom\"}],\"gas\":\"200000\"},\"memo\":\"Sent from Ledger\",\"msgs\":[{\"type\":\"cosmos-sdk/MsgVote\",\"value\":{\"option\":\"YES\",\"proposal_id\":\"123\",\"voter\":\"cosmos102hty0jv2s29lyc4u0tv97z9v298e24t3vwtpl\"}}],\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 13]
```diff
@@ -193,9 +226,9 @@ TEST(CosmosTransaction, ParseRawMsgVoteTransaction) {
     EXPECT_EQ(tx->serialize(), strTx);
 }
 
-TEST(CosmosTransaction, ParseRawMsgDepositTransaction) {
+TEST_F(CosmosTransactionTest, ParseRawMsgDepositTransaction) {
     auto strTx = "{\"account_number\":\"6571\",\"chain_id\":\"cosmoshub-3\",\"fee\":{\"amount\":[{\"amount\":\"5000\",\"denom\":\"uatom\"}],\"gas\":\"200000\"},\"memo\":\"Sent from Ledger\",\"msgs\":[{\"type\":\"cosmos-sdk/MsgDeposit\",\"value\":{\"amount\":[{\"amount\":\"1000000\",\"denom\":\"uatom\"}],\"depositor\":\"cosmos102hty0jv2s29lyc4u0tv97z9v298e24t3vwtpl\",\"proposal_id\":\"123\"}}],\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 14]
```diff
@@ -210,9 +243,9 @@ TEST(CosmosTransaction, ParseRawMsgDepositTransaction) {
     EXPECT_EQ(tx->serialize(), strTx);
 }
 
-TEST(CosmosTransaction, ParseRawMsgWithdrawDelegationRewardTransaction) {
+TEST_F(CosmosTransactionTest, ParseRawMsgWithdrawDelegationRewardTransaction) {
     auto strTx = "{\"account_number\":\"6571\",\"chain_id\":\"cosmoshub-3\",\"fee\":{\"amount\":[{\"amount\":\"5000\",\"denom\":\"uatom\"}],\"gas\":\"200000\"},\"memo\":\"Sent from Ledger\",\"msgs\":[{\"type\":\"cosmos-sdk/MsgWithdrawDelegationReward\",\"value\":{\"delegator_address\":\"cosmos102hty0jv2s29lyc4u0tv97z9v298e24t3vwtpl\",\"validator_address\":\"cosmosvaloper1grgelyng2v6v3t8z87wu3sxgt9m5s03xfytvz7\"}}],\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 15]
```diff
@@ -238,7 +271,7 @@ TEST(CosmosTransaction, ParseRawMsgMultiSendTransaction) {
         ",\"outputs\":[{\"address\":\"cosmos1da6hgur4wsmpnjyg\",\"coins\":[{\"amount\":\"10\",\"denom\":\"atom\"}]}]}}"
         "],"
         "\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 16]
```diff
@@ -276,7 +309,7 @@ TEST(CosmosTransaction, ParseRawMsgCreateValidatorTransaction) {
         "\"value\":{\"amount\":\"1059860\",\"denom\":\"uatom\"}"
         "}}],"
         "\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 17]
```diff
@@ -315,7 +348,7 @@ TEST(CosmosTransaction, ParseRawMsgEditValidatorTransaction) {
         "\"validator_address\":\"cosmostest\""
         "}}],"
         "\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 18]
```diff
@@ -345,7 +378,7 @@ TEST(CosmosTransaction, ParseRawMsgSetWithdrawAddressTransaction) {
         "\"withdraw_address\":\"cosmos1erfdsa\""
         "}}],"
         "\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 19]
```diff
@@ -370,7 +403,7 @@ TEST(CosmosTransaction, ParseRawMsgWithdrawDelegatorRewardsTransaction) {
         "\"validator_address\":\"cosmosvaloperwdfae\""
         "}}],"
         "\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 20]
```diff
@@ -394,7 +427,7 @@ TEST(CosmosTransaction, ParseRawMsgWithdrawValidatorCommissionTransaction) {
         "\"validator_address\":\"cosmosvaloper1234567890\""
         "}}],"
         "\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

File: core/test/cosmos/transactions_test.cpp [hunk 21]
```diff
@@ -417,7 +450,7 @@ TEST(CosmosTransaction, ParseRawMsgUnjailTransaction) {
         "\"validator_address\":\"cosmosvaloper1dalton\""
         "}}],"
         "\"sequence\":\"0\"}";
-    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(ledger::core::currencies::ATOM, strTx);
+    auto tx = api::CosmosLikeTransactionBuilder::parseRawUnsignedTransaction(currencies::ATOM, strTx);
```
Comment (gagbo) [inline core/test/cosmos/transactions_test.cpp]: If those are const objects, better mark them const too

**Review by gagbo — APPROVED** [review]
Looks good to me :) I think there's an easy way out for the `api::to_string` sending uppercase issue, otherwise it's cool !

At some point we'll start making tests and code const correct too, better start now if possible (I think the big libcore PR will also like better constness, but we forgot a lot of them up to this point so it's really going to be baby steps to get this going)

**Comment by ghost** [comment]
> Looks good to me :) I think there's an easy way out for the `api::to_string` sending uppercase issue, otherwise it's cool !
> 
> At some point we'll start making tests and code const correct too, better start now if possible (I think the big libcore PR will also like better constness, but we forgot a lot of them up to this point so it's really going to be baby steps to get this going)

All done, everywhere possible. Thanks for the review!

**Pull request merged**